src/phpDocumentor/ConfigurationFactory.php

```diff
@@ -0,0 +1,138 @@
+<?php
+
+namespace phpDocumentor;
+
+final class ConfigurationFactory
+{
+    /**
+     * @var Uri
+     */
+    private $uri;
+
+    /**
+     * @var \SimpleXMLElement
+     */
+    private $xml;
+
+    /**
+     * @var string
+     */
+    private $schemaPath;
+
+    /**
+     * @var bool
+     */
+    private $validateUri;
+
+    /**
+     * @param Uri    $uri
+     * @param string $schemaPath
+     */
+    public function __construct(Uri $uri, $schemaPath = '')
+    {
+        if ($schemaPath === '') {
+            $schemaPath = __DIR__ . '/../../data/xsd/phpdoc.xsd';
+        }
+
+        $this->replaceLocation($uri);
+        $this->schemaPath = $schemaPath;
+    }
+
+    /**
+     * Replaces the location of the configuration file if it is different.
+     *
+     * @param Uri $uri
+     */
+    public function replaceLocation(Uri $uri)
+    {
+        if ($this->uri !== $uri) {
+            $this->validateUri = true;
```

Comment:
Instead of tracking the cache using a boolean you can also clear the xml property and force the loading method to reload from the current uri. This will reduce the number of properties and hence reduce the complexity